packages/core/src/engine/Autodiff.ts

```diff
@@ -258,41 +265,94 @@ const children = (x: VarAD): Child[] => {
     }
     case "Unary": {
       return [
-        { child: x.param, name: undefined, sensitivity: unarySensitivity(x) },
+        {
+          child: x.param,
+          name: undefined,
+          sensitivity: [[unarySensitivity(x)]],
+        },
       ];
     }
     case "Binary": {
       const { left, right } = binarySensitivities(x);
       return [
-        { child: x.left, name: "left", sensitivity: left },
-        { child: x.right, name: "right", sensitivity: right },
+        { child: x.left, name: "left", sensitivity: [[left]] },
+        { child: x.right, name: "right", sensitivity: [[right]] },
+      ];
+    }
+    case "Comp":
+    case "Logic": {
+      return [
+        { child: x.left, name: "left", sensitivity: [] },
+        { child: x.right, name: "right", sensitivity: [] },
       ];
     }
     case "Ternary": {
       return [
-        { child: x.cond, name: "cond", sensitivity: undefined },
-        { child: x.then, name: "then", sensitivity: ifCond(x.cond, 1, 0) },
-        { child: x.els, name: "els", sensitivity: ifCond(x.cond, 0, 1) },
+        { child: x.cond, name: "cond", sensitivity: [[]] },
+        { child: x.then, name: "then", sensitivity: [[ifCond(x.cond, 1, 0)]] },
+        { child: x.els, name: "els", sensitivity: [[ifCond(x.cond, 0, 1)]] },
       ];
     }
     case "Nary": {
       return x.params.map((child, i) => {
         const c = { child, name: indexToNaryEdge(i) };
         switch (x.op) {
           case "addN": {
-            return { ...c, sensitivity: 1 };
+            return { ...c, sensitivity: [[1]] };
           }
           case "maxN": {
-            return { ...c, sensitivity: ifCond(lt(child, x), 0, 1) };
+            return { ...c, sensitivity: [[ifCond(lt(child, x), 0, 1)]] };
           }
           case "minN": {
-            return { ...c, sensitivity: ifCond(gt(child, x), 0, 1) };
+            return { ...c, sensitivity: [[ifCond(gt(child, x), 0, 1)]] };
           }
         }
       });
     }
+    case "PolyRoots": {
+      // https://www.skewray.com/articles/how-do-the-roots-of-a-polynomial-depend-on-the-coefficients
+
+      const n = x.coeffs.length;
+      const derivCoeffs: VarAD[] = x.coeffs.map((c, i) => mul(i, c));
+      derivCoeffs.shift();
+      // the polynomial is assumed monic, so `x.coeffs` doesn't include the
+      // coefficient 1 on the highest-degree term
+      derivCoeffs.push(n);
+
+      const sensitivities: VarAD[][] = x.coeffs.map((_, index) => {
+        const t: VarAD = { tag: "Index", index, vec: x }; // a root
+
+        let power: VarAD = 1;
+        const powers = [];
+        for (let i = 0; i < n; i++) {
+          powers.push(power);
+          power = mul(power, t);
+        }
+
+        const minusDerivative = neg(
+          addN(zip2(derivCoeffs, powers).map(([c, p]) => mul(c, p)))
```

Comment:
Ah good point, we actually already do have a dot product!

https://github.com/penrose/penrose/blob/32623914ee70b3aa267189a6bcee42a18c044574/packages/core/src/engine/Autodiff.ts#L645-L656

But the current implementation uses `reduce` and `add` rather than `addN` so it generates a larger graph.